exla/lib/exla/defn.ex

```diff
@@ -56,69 +56,227 @@ defmodule EXLA.Defn do
+    keep_on_device? = Keyword.get(run_options, :keep_on_device, false)
+    run_options = Keyword.put(run_options, :keep_on_device, true)
+
+    # The input vars should not be converted to buffers as they come from infeed
+    input_vars = Nx.Defn.Tree.flatten_list([input])
+    input_shape = EXLA.Shape.make_tuple_shape(Enum.map(input_vars, &nx_to_shape/1))
+    acc_vars = Nx.Defn.Tree.flatten_list([acc])
+    split_fun = &split_stream(&1, &2, length(input_vars), length(acc_vars))
+
+    comp_fun =
+      &to_stream_computation(key, input_vars, input_shape, acc_vars, &1, &2, compile_options)
+
+    {buffers, {:tuple, [output, acc_outputs]}, {executable, output_shape}} =
+      compile({:stream, key}, vars, fun, compile_options, split_fun, comp_fun)
+
+    # TODO: Add synchronization
+    # TODO: Should we transfer the buffers only before execution?
+    result = EXLA.Executable.run(executable, buffers, run_options)
+
+    %EXLA.Defn.Stream{
+      send: input,
+      send_shape: input_shape,
+      recv: output,
+      recv_shape: output_shape,
+      executable: executable,
+      done: fn -> buffer_to_nx(result, acc_outputs) end,
+      keep_on_device: keep_on_device?
+    }
+  end
+
+  defp split_stream(vars, used, input_length, acc_length) do
+    # Remove inputs from used buffers and include all accumulator entries.
+    total = input_length + acc_length
+
+    used =
+      Enum.to_list(input_length..(input_length + acc_length - 1)) ++
+        Enum.drop_while(used, &(&1 < total))
+
+    {Enum.take(vars, input_length), used}
+  end
+
+  defp to_stream_computation(key, input_vars, input_shape, acc_vars, expr, used_shapes, options) do
+    inspected_key = inspect(key)
+    builder = EXLA.Builder.new(inspected_key)
+
+    # Drop all accumulator entries from used_shapes as we will handle it separately.
+    used_shapes = Enum.drop(used_shapes, length(acc_vars))
+
+    # The stream loop will be a three element tuple:
+    #
+    #   The result of calling infeed.
+    #   The looping accumulator.
+    #   The looping constants.
+    #
+    # The input will be read as part of the infeed.
+    acc_shapes = Enum.map(acc_vars, &nx_to_shape/1)
+    acc_shape = EXLA.Shape.make_tuple_shape(acc_shapes)
+    constant_shape = EXLA.Shape.make_tuple_shape(Enum.map(used_shapes, &elem(&1, 1)))
+
+    flag_shape = EXLA.Shape.make_shape({:pred, 8}, {})
+    token_shape = EXLA.Shape.make_token_shape()
+    infeed_shape = EXLA.Shape.make_tuple_shape([flag_shape, token_shape])
+
+    arg_shape = EXLA.Shape.make_tuple_shape([infeed_shape, acc_shape, constant_shape])
+
+    cond_b = EXLA.Builder.new(builder, "while-cond-" <> inspected_key)
+    param = EXLA.Op.parameter(cond_b, 0, arg_shape, "arg")
+    infeed = EXLA.Op.get_tuple_element(param, 0)
+    flag = EXLA.Op.get_tuple_element(infeed, 0)
+    cond_op = EXLA.Op.equal(flag, EXLA.Op.constant_r0(cond_b, 1, {:pred, 8}))
+    cond = EXLA.Builder.build(cond_op)
```

Comment:
```suggestion
    predicate = EXLA.Builder.build(cond_op)
```

To avoid using an stdlib function name here

Reply:
Done!